lib/path-design.md

```diff
@@ -0,0 +1,177 @@
+  - :heavy_minus_sign: But only if you give it an argument of `.`. If you give it the argument `some-directory`, it won't prefix that
+- :heavy_minus_sign: `realpath --relative-to` doesn't output `./`'s
+  - :heavy_plus_sign: We don't need to return the same result though
+
+</details>
+
+### Representation of the current directory
+[curdir]: #representation-of-the-current-directory
+
+Context: The current directory can be represented with `.` or `./` or `./.`
+
+Decision: It should be `./.`
+
+<details>
+<summary>Arguments</summary>
+
+- :heavy_plus_sign: `./` would be inconsistent with [the decision to not have trailing slashes](#trailing-slashes)
+- :heavy_minus_sign: `.` is how `realpath` normalises paths
+- :heavy_plus_sign: `.` can be interpreted as a shell command (it's a builtin for sourcing files in bash and zsh)
+- :heavy_plus_sign: `.` would be the only path without a `/` and therefore not a valid Nix path in expressions
+- :heavy_minus_sign: `./.` is rather long
+  - :heavy_minus_sign: We don't require users to type this though, it's mainly just used as a library output.
+    As inputs all three variants are supported for relative paths (and we can't do anything about absolute paths)
+- :heavy_minus_sign: `builtins.dirOf "foo" == "."`, so `.` would be consistent with that
+- :heavy_plus_sign: `./.` is consistent with the [decision to have leading `./`](#leading-dots)
+
+</details>
+
+### Relative path representation
+[relrepr]: #relative-path-representation
+
+Context: Relative paths can be represented as a string, a list with all the components like `[ "foo" "bar" ]` for `foo/bar`, or with an attribute set like `{ type = "relative-path"; components = [ "foo" "bar" ]; }`
+
+Decision: Paths are represented as strings
+
+<details>
+<summary>Arguments</summary>
+
+- :heavy_plus_sign: It's simpler for the end user, as one doesn't need to make sure the path is in a string representation before it can be used
+  - :heavy_plus_sign: Also `concatStringsSep "/"` might be used to turn a relative list path value into a string, which then breaks for `[]`
+- :heavy_plus_sign: It doesn't encourage people to do their own path processing and instead use the library
+  E.g. With lists it would be very easy to just use `lib.lists.init` to get the parent directory, but then it breaks for `.`, represented as `[ ]`
+- :heavy_plus_sign: `+` is convenient and doesn't work on lists and attribute sets
+  - :heavy_minus_sign: Shouldn't use `+` anyways, we export safer functions for path manipulation
+
+</details>
+
+### Parents
+[parents]: #parents
+
+Context: Relative paths can have `..` components, which refer to the parent directory
+
+Decision: `..` path components in relative paths are not supported, nor as inputs nor as outputs.
+
+<details>
+<summary>Arguments</summary>
+
+- :heavy_plus_sign: It requires resolving symlinks to have proper behavior, since e.g. `foo/..` would not be the same as `.` if `foo` is a symlink.
```

Comment:
```suggestion
- :heavy_plus_sign: If we'd want symlinks to behave according to the "physical" interpretation, it requires resolving symlinks, since e.g. `foo/..` would not be the same as `.` if `foo` is a symlink.
  - :heavy_plus_sign: The "logical" interpretation is also valid, and is used by some software. It is simpler and therefore perhaps more secure.
  - :heavy_plus_sign: Mixing both models can lead to surprises.
```

Reply:
@roberth I suppose you meant this?

```suggestion
- :heavy_plus_sign: If we wanted symlinks to behave according to the "physical" interpretation, it would require resolving symlinks, since e.g. `foo/..` would not be the same as `.` if `foo` is a symlink.
  - :heavy_minus_sign: The "logical" interpretation is also valid, and is used by some software. It is simpler, and not using symlinks at all is safer.
  - :heavy_plus_sign: Mixing both models can lead to surprises.
```

Reply:
I inlined the explanations of those terms here for simplicity.